New plot_coupling_map(), like plot_gate_map() but without a backend

haggaila opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
The function `plot_gate_map()` currently requires a `backend` argument, and plots a hard-coded device map based on the number of qubits in the backend. Without breaking its interface, I moved most of its code into a second top-level function `plot_gate_map_data()` that takes instead of a backend three arguments - the number of qubits, `mpl_data` that describes the qubit coordinates on a planar grid, and `cmap` that gives the coupling map (these were all local variables of `plot_gate_map()` used in the implementation). `plot_gate_map()` now invokes `plot_gate_map_data()` with arguments appropriate for the backend, and in addition users can now call `plot_gate_map_data()` directly to plot device coupling maps in an arbitrary way, a very useful feature for simulations of qubit dynamics, such as when using qiskit-dynamics.


### Details and comments

nkanazawa1989 commented on qiskit/visualization/gate_map.py:
```diff
@@ -349,6 +335,82 @@ def plot_gate_map(
+def plot_gate_map_data(
+    num_qubits,
+    mpl_data,
+    cmap,
```
In many matplotlib example code we use `cmap` for the color map object and thus this is bit confusing. Can you explicitly say coupling_map? `mpl_data` is also not clear to me. Perhaps `qubit_coordinates`?

haggaila replied:
I kept the names as in the original implementation - sure I'll change according to your suggestion.

nkanazawa1989 reviewed (commented):
The change seems to be reasonable. However I would append `cmap`, `mpl_data`, `num_qubit` to the end of arguments and make `backend` optional and add  check something like

```python
if not backend and not all(cmap, mpl_data, num_qubits):   # I guess this doesn't work as-is
   raise some error
```

instead of adding new function (to reduce maintenance overhead). It's up to you.

nkanazawa1989 reviewed (approved):
LGTM. Could you please write reno?
https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#adding-a-new-release-note

nkanazawa1989 commented on releasenotes/notes/plot_gate_map-new-parameters-deb973b1bf0ad92f.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    The function ``qiskit.visualization.plot_gate_mpa()`` has been extended with new parameters.
+    If the backend argument is passed, the corresponding device's gate map will be plotted.
+    If the backend argument is None, three new parameters must be passed; ``num_qubit``,
+    ``qubit_coordinates``, and ``coupling_map``, which define a gate map to be plotted.
+
+
```
```suggestion
    The function :func:`~qiskit.visualization.plot_gate_map()` has been extended with new parameters.
    If the ``backend`` argument is passed, the corresponding device's gate map will be plotted.
    If the ``backend`` argument is ``None``, three new parameters must be passed; ``num_qubit``,
    ``qubit_coordinates``, and ``coupling_map``, which define a gate map to be plotted.

```

nkanazawa1989 commented on releasenotes/notes/plot_gate_map-new-parameters-deb973b1bf0ad92f.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+features:
```
```suggestion
upgrade:
```

nkanazawa1989 commented on qiskit/visualization/gate_map.py:
```diff
@@ -36,11 +36,14 @@ def plot_gate_map(
     Args:
-        backend (BaseBackend): A backend instance. If set to None (the default), then the three
-            arguments num_qubits, mpl_data, and cmap must not be None.
+        backend (BaseBackend): If set to a backend instance, it will be used to plot the device gate
+            map, and the three arguments num_qubits, mpl_data, and cmap must be None. Otherwise, if
+            backend is set to None, the aforementioned three arguments must be passed.
```
```suggestion
        backend (BaseBackend): If set to a backend instance, it will be used to plot the device gate
            map, and the three arguments ``num_qubits``, ``qubit_coordinates``, 
            and ``coupling_map`` must be None. Otherwise, if ``backend`` is set to None, 
            the aforementioned three arguments must be passed.
```

1ucian0 commented:
I actually like better the original idea, where `plot_gate_map` calls `plot_gate_map_data` and the user is free to call the second one if needed. I think overloading `plot_gate_map`'s parameter logic suggested by @nkanazawa1989 is not nice. These two functions have significantly different use cases. My two cents.

1ucian0 reviewed (changes requested):
My main comment is [here](https://github.com/Qiskit/qiskit-terra/pull/6751#issuecomment-890964698). However, this request-change review is to block the merge and ask for a test :)

If you decide to continue this approach, can you extend `test/python/visualization/test_gate_map.py` with this new API?

Thanks!

haggaila commented:
@1ucian0 I apologize for the long delay before responding. I agree that each option (new function or new parameters) has a bit of different advantages/disadvantages, but eventually I think that adding new parameters to the existing function is a bit favorable in terms of discoverability of the new functionality (which might go unnoticed when introducing a new function).
I pushed a new test function and a file saved from it.

haggaila commented:
@1ucian0, after discussing also with @nkanazawa1989, I changed the interface back to the original one for ``plot_gate_map()``, and added a new function ``plot_coupling_map()`` that works without a backend. I fixed the test function, but not the release notes - waiting for confirmation on the new code. Thanks.

1ucian0 reviewed (dismissed):
Thanks!
